### ISSUE TITLE
How to do a simple clone operation in basic.p4

### ISSUE THREAD

**Issue by Blandine11** [p4lang/behavioral-model#851]
Hello,
I begin using P4 language and I try to performs some basic exercices on github.
What I want to do  is to performs clone operation in basic.p4 (in p4 beginers tutorial on github) solution's file to send a clone packet to port 2 when a packet is sent to 10.0.3.3. I use bmv2 simple_switch.
I created a new action (clone_action) and I call clone function like this :
action clone_action ()
{ 
        clone(CloneType.I2E, 1024).
} 

 This action is called on table entries to do clone operation of packet.
In sxx-commands.txt I added the clone session id (1024) configuration to send the clone packet to port number 2  like this : 
mirroring_mapping_add(1024, 2).
But it does not work. 
So, I want to know where I must to configure the mapping of clone session id and the destination port? If it's not in the sxx-commands.txt.
The table in which I call  clone_action is the following : 
control myIngress(....)
{
table clone_tab{
key = { hdr.ipv4.dstAddr: exact;}
actions = { clone_action;...}
size = 1024;
}
apply{
        if(hdr.ipv4.isValid())
{
         clone_table.apply();
         ...
}
...
}
In the P4  program this table is called  clone_tab. My s1-commands.txt is the following :
 
mirroring_mapping_add(1024,2) // mapping clone id 1024 to port 2
table_set_default ipv4_lpm drop
table_add clone_tab clone_action 10.0.3.3 
 To summarize, I want to know where and how to configure clone session id while using bmv2 simple_switch in order to be able to use it in p4 program.

Regards

**Comment by jafingerhut** [comment]
I suspect that the portion that is not working is this: "In sxx-commands.txt I added the clone session id (1024) configuration".

By that I mean that I know of two different APIs by which the simple_switch process, that is executing your P4 code, can accept and correctly process a request to make such a change, using a Thrift API, or using the P4Runtime API.

What I do not know enough to answer your question is how the p4lang/tutorials repository code processes the contents of the files named sxx-commands.txt to turn them into Thrift and/or P4Runtime API requests to the simple_switch process, and I suspect it is that part of the puzzle that is missing in your case.

Thus your question is likely to be specific to code in the p4lang/tutorials repository, not p4lang/behavioral-model, and _maybe_ creating an issue there will be more likely to get attention from people who know how the code in that p4lang/tutorials repository works.

For example, it is possible (I do no know for sure), that given the code in the p4lang/tutorials repository right now, there is no way to do what you want no matter how you edit the sxx-commands.txt file.  It really depends upon the code that reads that file, and what it does, and I do not know which code that is.

**Comment by antoninbas** [comment]
Where does the idea for adding `mirroring_mapping_add(1024,2)` come from? Based on what you are writing it seems that the `sX-commands.txt` file is just a list of simple_switch CLI commands, in which case the appropriate command should be: `mirroring_add <mirror_id> <egress_port>` (see https://github.com/p4lang/behavioral-model/blob/master/targets/simple_switch/sswitch_CLI.py#L86).

This command is not documented in https://github.com/p4lang/behavioral-model/blob/master/docs/runtime_CLI.md unfortunately, because it is specific to the simple_switch bmv2 target. You can run the `simple_swith_CLI` tool directly however, and look at the full list of commands by typing `help`.

**Comment by Blandine11** [comment]
Hello Guys,
Thanks you for your quick replies.
@jafingerhut, You saw right. The part of the puzzle that was missing is the wrong command I used to configure clone session id and egress port (i.e. _mirroring_mapping_add(1024,2)_).
@antoninbas, as you suggest, I used  the command _mirroring_add 1024 2__  and it works!!

Regards
